lib/srv/desktop/rdp/rdpclient/src/scard.rs

```diff
@@ -0,0 +1,1337 @@
+            "invalid NDR pointer value {:#010X}, expected {:#010X}",
+            ptr, expect_ptr
+        )))
+    } else {
+        Ok(ptr)
+    }
+}
+
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+struct ListReaders_Call {
+    context: Context,
+    groups_length: u32,
+    groups: Vec<String>,
+    readers_is_null: bool,
+    readers_size: u32,
+}
+
+impl ListReaders_Call {
+    fn decode(payload: &mut Payload) -> RdpResult<Self> {
+        let _header = RPCEStreamHeader::decode(payload)?;
+        let _header = RPCETypeHeader::decode(payload)?;
+
+        let mut index = 0;
+        let mut context = Context::decode_ptr(payload, &mut index)?;
+        let groups_ptr_length = payload.read_u32::<LittleEndian>()?;
+        let groups_ptr = decode_ptr(payload, &mut index)?;
+        let readers_is_null = (payload.read_u32::<LittleEndian>()?) == 0x00000001;
+        let readers_size = payload.read_u32::<LittleEndian>()?;
+
+        context.decode_value(payload)?;
+
+        if groups_ptr == 0 {
+            return Ok(Self {
+                context,
+                groups_length: 0,
+                groups: Vec::new(),
+                readers_is_null,
+                readers_size,
+            });
+        }
+        let (groups_length, groups) = decode_multistring_unicode(payload)?;
+        if groups_length != groups_ptr_length {
+            Err(invalid_data_error(
+                "mismatched reader groups length in NDR pointer and value",
+            ))
+        } else {
+            Ok(Self {
+                context,
+                groups_length,
+                groups,
+                readers_is_null,
+                readers_size,
+            })
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+struct ListReaders_Return {
+    return_code: ReturnCode,
+    readers: Vec<String>,
+}
+
+impl ListReaders_Return {
+    fn new(return_code: ReturnCode, readers: Vec<String>) -> Self {
+        Self {
+            return_code,
+            readers,
+        }
+    }
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.write_i32::<LittleEndian>(self.return_code.to_i32().unwrap())?;
+        let readers = encode_multistring_unicode(&self.readers)?;
+        let mut index = 0;
+        let ptr = encode_ptr(readers.length() as u32, &mut index)?;
+        w.extend_from_slice(&ptr);
+
+        w.write_u32::<LittleEndian>(readers.length() as u32)?;
+        w.extend_from_slice(&readers);
+
+        Ok(encode_response(encode_struct(w)?)?)
+    }
+}
+
+fn decode_multistring_unicode(payload: &mut Payload) -> RdpResult<(u32, Vec<String>)> {
+    let len = payload.read_u32::<LittleEndian>()?;
+    let mut items = vec![];
+    let mut buf = vec![];
+    for _i in 0..(len / 2) {
+        let c = payload.read_u16::<LittleEndian>()?;
+        if c == 0 {
+            if buf.len() > 0 {
+                items.push(
+                    decode_utf16(buf.iter().cloned())
```

Comment:
Can you use `copied()` instead?

Reply:
So `cloned` delegates to the `Clone` trait while `Copy` is a special marker trait. `copied` delegates to a compiler intrinsic which usually generates better code for some reason.